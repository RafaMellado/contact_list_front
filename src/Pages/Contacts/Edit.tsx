import { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactsService from "../../Services/ContactsService";
import {
  Contact,
  ContactRequestBody,
  ContactRequestError,
} from "../../Services/Interfaces/Contact";
import { ContactForm } from "./Components/Form";

export function ContactEdit() {
  const [contact, setContact] = useState<Contact>();

  const { id } = useParams();
  const { t } = useTranslation();
  const navigate = useNavigate();
  const [errors, setErrors] = useState<ContactRequestError>();

  const TRANSLATIONS: string = "contactsEdit";

  const backToList = () => {
    navigate(`/contact-book/show/${contact?.contact_book_id}`);
  };

  const editContact = async (data: Partial<ContactRequestBody>) => {
    const newData = {
      ...data,
      contact_book_id: Number(contact?.contact_book_id),
    };

    try {
      await ContactsService.update(Number(id), newData as ContactRequestBody);

      backToList();
    } catch (errors) {
      setErrors(errors as ContactRequestError);
    }
  };

  useEffect(() => {
    ContactsService.show(Number(id)).then((response) => setContact(response));
  }, [id]);

  return (
    <div className="d-flex flex-column align-items-center">
      <h2 className="mb-4">
        {t<string>(`${TRANSLATIONS}.title`, {
          name: `${contact?.givenname} ${contact?.surname}`,
        })}
      </h2>

      <ContactForm
        data={contact}
        translations={TRANSLATIONS}
        errors={errors}
        submitFn={editContact}
        backToListFn={backToList}
      />
    </div>
  );
}
