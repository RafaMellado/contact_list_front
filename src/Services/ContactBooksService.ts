import {
  ContactBook,
  ContactBookRequestBody,
  ContactBookWithContacts,
} from "./Interfaces/ContactBook";
import ContactBooksRepository from "./Repository/ContactBooksRepository";

class ContactBooksService {
  index(): Promise<ContactBook[]> {
    return ContactBooksRepository.index();
  }

  show(id: number): Promise<ContactBookWithContacts> {
    return ContactBooksRepository.show(id);
  }

  create(data: ContactBookRequestBody): Promise<ContactBook> {
    return ContactBooksRepository.create(data);
  }

  update(id: number, data: ContactBookRequestBody): Promise<ContactBook> {
    return ContactBooksRepository.update(id, data);
  }

  delete(id: number): Promise<Response> {
    return ContactBooksRepository.delete(id);
  }
}

export default new ContactBooksService();
