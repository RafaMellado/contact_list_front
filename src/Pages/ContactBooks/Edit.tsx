import { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate, useParams } from "react-router-dom";
import ContactBooksService from "../../Services/ContactBooksService";
import {
  ContactBook,
  ContactBookRequestBody,
  ContactBookRequestError,
} from "../../Services/Interfaces/ContactBook";
import { ContactBookForm } from "./Components/Form";

export function ContactBookEdit() {
  const [contactBook, setContactBook] = useState<ContactBook>();
  const [errors, setErrors] = useState<ContactBookRequestError>();

  const { id } = useParams();
  const { t } = useTranslation();
  const navigate = useNavigate();

  const TRANSLATIONS: string = "contactBooksEdit";

  const backToHome = () => {
    navigate("/home");
  };

  const editContactBook = async (data: ContactBookRequestBody) => {
    try {
      await ContactBooksService.update(Number(id), data);

      backToHome();
    } catch (errors) {
      setErrors(errors as ContactBookRequestError);
    }
  };

  const getContactBook = async () => {
    const response = await ContactBooksService.show(Number(id));

    setContactBook(response);
  };

  useEffect(() => {
    getContactBook();
  }, [id]);

  return (
    <div className="d-flex flex-column align-items-center">
      {contactBook?.name && (
        <h2 data-testid="contact-book-edit-title" className="mb-4">
          {t<string>(`${TRANSLATIONS}.title`, { name: contactBook.name })}
        </h2>
      )}
      <ContactBookForm
        translations={TRANSLATIONS}
        item={contactBook}
        submitFn={editContactBook}
        backFn={backToHome}
        errors={errors}
      />
    </div>
  );
}
