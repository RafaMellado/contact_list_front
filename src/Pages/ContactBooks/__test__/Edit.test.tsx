import { ContactBookEdit } from "../Edit";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import ContactBooksService from "../../../Services/ContactBooksService";

jest.mock("react-i18next", () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({ id: 1 }),
}));

jest.mock("../../../Services/ContactBooksService");

const mockedContactBookService = ContactBooksService as jest.Mocked<
  typeof ContactBooksService
>;

describe(`<${ContactBookEdit.name} />`, () => {
  const history = createMemoryHistory();

  const factoryComponent = async () => {
    const component = render(
      <Router location="/" navigator={history}>
        <ContactBookEdit />
      </Router>
    );

    await waitFor(() => {
      expect(
        component.getByTestId("contact-book-edit-title")
      ).toBeInTheDocument();
    });

    return component;
  };

  beforeEach(() => {
    mockedContactBookService.show.mockResolvedValue({
      id: 1,
      name: "test",
      user_id: 1,
    });
  });

  test("renders correctly the component", async () => {
    const component = await factoryComponent();

    expect(component.asFragment()).toMatchSnapshot();
  });

  test("click on backFn btn redirects correctly", async () => {
    const component = await factoryComponent();

    fireEvent.click(component.getByTestId("contact-book-back-btn"));

    expect(history.location.pathname).toBe("/home");
  });

  test("submit form calls editContactBook correctly", async () => {
    const component = await factoryComponent();

    fireEvent.submit(component.getByTestId("contact-book-form"));

    expect(mockedContactBookService.update).toHaveBeenCalledWith(1, {
      name: "test",
    });
  });

  test("submit form calls editContactBook correctly and set errors", async () => {
    mockedContactBookService.update.mockRejectedValue({
      errors: {
        name: [{ error: "name", value: "test" }],
      },
    });

    const component = await factoryComponent();

    fireEvent.submit(component.getByTestId("contact-book-form"));

    await waitFor(() => {
      expect(component.queryByTestId("error-box-name")).toBeTruthy();
    });

    expect(component.asFragment()).toMatchSnapshot("errors");
  });
});
