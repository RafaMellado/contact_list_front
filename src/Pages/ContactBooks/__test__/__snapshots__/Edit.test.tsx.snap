// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactBookEdit /> renders correctly the component 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column align-items-center"
  >
    <h2
      class="mb-4"
      data-testid="contact-book-edit-title"
    >
      contactBooksEdit.title
    </h2>
    <div
      class="form card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          data-testid="contact-book-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactBooksEdit.name
            </label>
            <input
              class="form-control"
              maxlength="20"
              minlength="4"
              placeholder="contactBooksEdit.name"
              required=""
              type="text"
              value="test"
            />
          </div>
          <div
            class="d-flex justify-content-between"
          >
            <button
              class="w-100 btn btn-secondary"
              data-testid="contact-book-back-btn"
              type="button"
            >
              contactBooksEdit.back
            </button>
            <button
              class="w-100 ms-2 btn btn-primary"
              type="submit"
            >
              contactBooksEdit.submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ContactBookEdit /> submit form calls editContactBook correctly and set errors: errors 1`] = `
<DocumentFragment>
  <div
    class="d-flex flex-column align-items-center"
  >
    <h2
      class="mb-4"
      data-testid="contact-book-edit-title"
    >
      contactBooksEdit.title
    </h2>
    <div
      class="form card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          data-testid="contact-book-form"
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              contactBooksEdit.name
            </label>
            <input
              class="form-control"
              maxlength="20"
              minlength="4"
              placeholder="contactBooksEdit.name"
              required=""
              type="text"
              value="test"
            />
            <div
              class="text-danger mt-2 mb-0"
              data-testid="error-box-name"
            >
              <span>
                errors.name
              </span>
            </div>
          </div>
          <div
            class="d-flex justify-content-between"
          >
            <button
              class="w-100 btn btn-secondary"
              data-testid="contact-book-back-btn"
              type="button"
            >
              contactBooksEdit.back
            </button>
            <button
              class="w-100 ms-2 btn btn-primary"
              type="submit"
            >
              contactBooksEdit.submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
